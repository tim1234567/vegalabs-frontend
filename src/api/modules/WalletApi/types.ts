import { AbstractConnector } from '@web3-wallets-kit/abstract-connector';
import { DefaultConnectionPayload } from '@web3-wallets-kit/core';

import { Network, WalletType } from 'domain/types';
import { NetworkID } from 'env';

import { ConnectWalletApi } from './ConnectWalletApi';
import { Web3WalletApi } from './Web3WalletApi';

export type WalletVersion = {
  version: string;
  isMobile: boolean;
  walletType?: 'metamask';
};

export interface WalletApiClass {
  readonly name: WalletType;
  readonly isMultiChain: false;
  readonly supportedChains: Network[];
  getConnector(network: Network): AbstractConnector<DefaultConnectionPayload> | null;
  getWalletVersion?(): Promise<WalletVersion | null>;
  watchAsset?(options: {
    address: string;
    decimals: number;
    symbol: string;
    image?: string;
  }): Promise<boolean>;
}

export interface MultiChainWalletApiClass extends Omit<WalletApiClass, 'isMultiChain'> {
  readonly isMultiChain: true;
  switchNetwork(chainId: NetworkID): Promise<void>;
  prepareWalletForTransaction(chainId: NetworkID): Promise<void>;
}

export type WalletApi = ConnectWalletApi | Web3WalletApi;
