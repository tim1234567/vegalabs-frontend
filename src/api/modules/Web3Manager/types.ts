import { Network, WalletType } from 'domain/types';

export interface StorageStateV1 {
  lastProvider: null | WalletType;
  lastNetwork: Network;
}
