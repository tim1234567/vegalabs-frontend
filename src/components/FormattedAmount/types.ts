import { Amount } from '@akropolis-web/primitives';
import { TooltipClassKey } from '@mui/material';

import { MaybeEither } from 'utils/either';

export type ElementsClasses = {
  root?: string;
  symbol?: string;
  tooltip?: Partial<Record<TooltipClassKey, string>>;
};

export type FormattedAmountProps = {
  sum: MaybeEither<Amount>;
  precision?: number;
  className?: string;
  classes?: ElementsClasses;
  hasSign?: boolean;
  withSI?: boolean;
  withInfoIcon?: boolean;
  symbolVariant?: 'none' | 'text';
  symbolSize?: 'small' | 'medium' | 'inherit';
  tooltip?: JSX.Element | 'none' | 'default';
  wrap?: 'wrap' | 'nowrap';
};
