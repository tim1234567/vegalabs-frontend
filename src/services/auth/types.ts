import { ConnectionStatus, ConnectResult } from '@web3-wallets-kit/core';
import { Observable, BehaviorSubject } from 'rxjs';

import { Network, WalletType } from 'domain/types';
import { Communication } from 'utils/react';
import { WalletApi } from 'api/modules/WalletApi';

interface AuthWeb3Manager {
  account$: Observable<string | null>;
  status$: BehaviorSubject<ConnectionStatus>;
  chainId$: BehaviorSubject<number | null>;
  connect: (
    wallet: WalletType,
    chainId: Network,
    isAutoReconnect: boolean,
  ) => Promise<ConnectResult>;
  disconnect: () => void;
  connectedWallet$: BehaviorSubject<WalletApi | null>;
}

export type ConnectionRequest = {
  toNetwork?: Network;
  toAccount?: string;
  onAccountWarningClose?: () => void;
  onNetworkWarningClose?: () => void;
  onAuthModalClose?: () => void;
};

// success – suitable account and network
// unsuitable-account – unsuitable account and any network
// unsuitable-network – suitable account and unsuitable network
export type ConnectingResult =
  | {
      status: 'success';
    }
  | {
      status: 'unsuitable-network';
      targetNetwork: Network;
      currentChainId: number;
    }
  | {
      status: 'unsuitable-account';
      targetAccount: string;
      currentAccount: string;
    };

export interface AuthContextType {
  web3Manager: AuthWeb3Manager;
  connectCommunication: Communication<
    (wallet: WalletType, network: Network) => Promise<ConnectingResult>
  >;
  openedModal: 'auth' | 'unsuitable-network-warning' | 'unsuitable-account-warning' | null;
  openModal: (options?: ConnectionRequest) => void;
  closeModal: () => void;
}
