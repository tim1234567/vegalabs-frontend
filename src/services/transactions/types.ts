import { TransactionReceipt } from 'web3-core';

import { RemoteData } from 'utils/remoteData';
import { Network } from 'domain/types';

export type TransactionPayload = {
  txStatusContent: {
    title: {
      long: string;
      short: string;
    };
    network: Network;
  };
};

export type TxStatus = {
  txHash: string | null;
  result: RemoteData<TransactionReceipt>;
};

const statuses = ['initial', 'pending', 'failed', 'cancelled', 'success'] as const;
export type Statuses = typeof statuses[number];
