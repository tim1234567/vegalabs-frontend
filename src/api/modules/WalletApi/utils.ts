import { AbstractConnector } from '@web3-wallets-kit/abstract-connector';
import { DefaultConnectionPayload } from '@web3-wallets-kit/core';

import { DeferredPromise } from 'utils/js';
import { Network, WalletType } from 'domain/types';
import { NetworkID, SWITCH_NETWORK_TIMEOUT } from 'env';

import { WalletApi } from './types';
import { ConnectWalletApi } from './ConnectWalletApi';
import { Web3WalletApi } from './Web3WalletApi';

export function makeConnectors<P extends AbstractConnector<DefaultConnectionPayload>>(
  networks: Network[],
  makeConnector: (network: Network) => P,
) {
  return networks.reduce(
    (acc, network) => ({ ...acc, [network]: makeConnector(network) }),
    {} as Partial<Record<Network, P>>,
  );
}

export function waitForNetworkChange(
  connector: AbstractConnector<DefaultConnectionPayload>,
  expectedChainId: NetworkID,
): Promise<void> {
  const result = new DeferredPromise<void>();
  const subscription = connector.subscribeChainId((chainId: number) => {
    if (chainId === expectedChainId) {
      clearTimeout(timeout);
      subscription.unsubscribe();
      result.resolve();
    }
  });

  const timeout = setTimeout(() => {
    subscription.unsubscribe();
    result.reject(`Failed to wait for the chainId ${expectedChainId}`);
  }, SWITCH_NETWORK_TIMEOUT);

  return result.promise;
}

export function makeWallets(): Record<WalletType, WalletApi> {
  return {
    connectWallet: new ConnectWalletApi(),
    web3: new Web3WalletApi(),
  };
}
