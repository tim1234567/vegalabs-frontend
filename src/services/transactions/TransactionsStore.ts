import { Observable } from 'rxjs';
import * as R from 'ramda';
import { Service } from 'typedi';

import { select, update, getValue, createStore } from 'core/store';

import { TransactionPayload, TxStatus } from './types';

type TransactionID = string;
type TxStatuses = Record<TransactionID, TxStatus | undefined>;

@Service()
class TransactionsStore {
  public transactionPayloads = createStore<Record<TransactionID, TransactionPayload | undefined>>(
    () => ({}),
  );

  public txStatuses = createStore<TxStatuses>(() => ({}));
  public inlineStatuses = createStore<TransactionID[]>(() => []);
  public globalStatuses = createStore<TransactionID[]>(() => []);

  public getTxStatusByID$(id: TransactionID): Observable<TxStatus | null> {
    return select(this.txStatuses, value => value[id] || null);
  }

  public getTransactionPayloadByID$(id: TransactionID) {
    return select(this.transactionPayloads, value => value[id] || null);
  }

  public addTxStatus(id: TransactionID, status: TxStatus): void {
    update(this.txStatuses, value =>
      R.mergeRight(value, {
        [id]: status,
      }),
    );
  }

  public updateTxStatus(id: TransactionID, status: Partial<TxStatus>): void {
    update(this.txStatuses, value => {
      const prevStatus = value[id];

      // ignore if status already cleaned
      if (!prevStatus) {
        console.warn(
          'transactions/TransactionsStore: impossible to update the tx status, previous tx status is not found',
        );
        return value;
      }

      return R.mergeRight(value, {
        [id]: R.mergeRight(prevStatus, status),
      });
    });
  }

  public addTransactionPayload(id: TransactionID, payload: TransactionPayload): void {
    update(this.transactionPayloads, value =>
      R.mergeRight(value, {
        [id]: payload,
      }),
    );
  }

  public switchStatusLocation(id: TransactionID): void {
    const currentLocation =
      (getValue(this.inlineStatuses).includes(id) && 'inline') ||
      (getValue(this.globalStatuses).includes(id) && 'global') ||
      null;

    currentLocation === 'global' && this.displayStatusInline(id);
    currentLocation === 'inline' && this.displayStatusGlobal(id);
  }

  public displayStatusInline(id: TransactionID): void {
    update(this.inlineStatuses, value => R.uniq([...value, id]));
    update(this.globalStatuses, value => R.without([id], value));
  }

  public displayStatusGlobal(id: TransactionID): void {
    update(this.globalStatuses, value => R.uniq([...value, id]));
    update(this.inlineStatuses, value => R.without([id], value));
  }

  public hideStatus(id: TransactionID): void {
    this.hideStatuses([id]);
  }

  public hideStatuses(ids: TransactionID[]): void {
    update(this.globalStatuses, value => R.without(ids, value));
    update(this.inlineStatuses, value => R.without(ids, value));
    this.cleanState();
  }

  private cleanState() {
    const usefulIds = R.uniq([getValue(this.globalStatuses), getValue(this.inlineStatuses)].flat());

    const unusedIds = R.without(
      usefulIds,
      R.uniq(
        [
          Object.keys(getValue(this.transactionPayloads)),
          Object.keys(getValue(this.txStatuses)),
        ].flat(),
      ),
    );

    if (!unusedIds.length) {
      return;
    }

    update(this.transactionPayloads, value => R.omit(unusedIds, value));
    update(this.txStatuses, value => R.omit(unusedIds, value));
  }
}

export { TransactionsStore };
