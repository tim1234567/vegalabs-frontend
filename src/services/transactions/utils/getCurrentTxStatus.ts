import { isFailure, isSuccess } from 'utils/remoteData';

import { TxStatus, Statuses } from '../types';

export function getCurrentTxStatus(status: TxStatus): Statuses {
  const { result, txHash } = status;
  if (!txHash && isFailure(result)) return 'cancelled';
  if (txHash && isFailure(result)) return 'failed';
  if (txHash && isSuccess(result)) return 'success';
  if (txHash) return 'pending';
  return 'initial';
}
