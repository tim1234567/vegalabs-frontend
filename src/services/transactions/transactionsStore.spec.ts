import { TransactionReceipt } from 'web3-core';

import { getValue } from 'core/store';
import { loading, success } from 'utils/remoteData';

import { TransactionsStore } from './TransactionsStore';
import { TransactionPayload } from './types';

const receiptMock: TransactionReceipt = 'ReceiptMock' as unknown as TransactionReceipt;

describe('addTxStatus', () => {
  let transactionsStore: TransactionsStore;

  beforeEach(() => {
    transactionsStore = new TransactionsStore();
    transactionsStore.addTxStatus('1', { txHash: '1', result: loading });
    transactionsStore.addTxStatus('2', { txHash: '2', result: loading });
  });

  it('should create new status record and keep old statuses', () => {
    const prevValue = getValue(transactionsStore.txStatuses);
    const args = ['3', { txHash: '3', result: loading }] as const;

    transactionsStore.addTxStatus(...args);
    expect({
      prevValue,
      args,
      result: getValue(transactionsStore.txStatuses),
    }).toMatchSnapshot();
  });

  it('should replace status for existing transaction ID', () => {
    const prevValue = getValue(transactionsStore.txStatuses);
    const args = ['1', { txHash: '111', result: success(receiptMock) }] as const;

    transactionsStore.addTxStatus(...args);
    expect({
      prevValue,
      args,
      result: getValue(transactionsStore.txStatuses),
    }).toMatchSnapshot();
  });
});

describe('updateTxStatus', () => {
  let transactionsStore: TransactionsStore;

  beforeEach(() => {
    transactionsStore = new TransactionsStore();
    transactionsStore.addTxStatus('1', { txHash: '1', result: loading });
    transactionsStore.addTxStatus('2', { txHash: '2', result: loading });
  });

  it('should partially update status for existing transaction ID', () => {
    const prevValue = getValue(transactionsStore.txStatuses);
    const args = ['1', { txHash: '111' }] as const;

    transactionsStore.updateTxStatus(...args);
    expect({
      prevValue,
      args,
      result: getValue(transactionsStore.txStatuses),
    }).toMatchSnapshot();
  });

  it('should not change statuses if they do not contain the transaction ID', () => {
    const prevValue = getValue(transactionsStore.txStatuses);
    const args = ['3', { txHash: '3' }] as const;

    transactionsStore.updateTxStatus(...args);
    expect({
      prevValue,
      args,
      result: getValue(transactionsStore.txStatuses),
    }).toMatchSnapshot();
  });
});

describe('addTransactionPayload', () => {
  let transactionsStore: TransactionsStore;

  beforeEach(() => {
    transactionsStore = new TransactionsStore();
    transactionsStore.addTransactionPayload('1', 'txPayload' as unknown as TransactionPayload);
    transactionsStore.addTransactionPayload('2', 'txPayload' as unknown as TransactionPayload);
  });

  it('should create new payload record and keep old payloads', () => {
    const prevValue = getValue(transactionsStore.transactionPayloads);
    const args = ['3', 'txPayload' as unknown as TransactionPayload] as const;

    transactionsStore.addTransactionPayload(...args);
    expect({
      prevValue,
      args,
      result: getValue(transactionsStore.transactionPayloads),
    }).toMatchSnapshot();
  });

  it('should replace payload for existing transaction ID', () => {
    const prevValue = getValue(transactionsStore.transactionPayloads);
    const args = [
      '1',
      {
        txStatusContent: { title: { long: '111', short: '1' }, network: 'polygon' },
      },
    ] as const;

    transactionsStore.addTransactionPayload(...args);
    expect({
      prevValue,
      args,
      result: getValue(transactionsStore.transactionPayloads),
    }).toMatchSnapshot();
  });
});

describe('switchStatusLocation', () => {
  let transactionsStore: TransactionsStore;

  beforeEach(() => {
    transactionsStore = new TransactionsStore();
    transactionsStore.displayStatusInline('1');
    transactionsStore.displayStatusInline('2');
    transactionsStore.displayStatusGlobal('3');
    transactionsStore.displayStatusGlobal('4');
  });

  it('should keep old statuses if switched status is not found', () => {
    const prevValue = {
      inlineStatuses: getValue(transactionsStore.inlineStatuses),
      globalStatuses: getValue(transactionsStore.globalStatuses),
    };
    const args = '5';
    transactionsStore.switchStatusLocation(args);

    expect({
      prevValue,
      args,
      result: {
        inlineStatuses: getValue(transactionsStore.inlineStatuses),
        globalStatuses: getValue(transactionsStore.globalStatuses),
      },
    }).toMatchSnapshot();
  });

  it('should move status from inline to global', () => {
    const prevValue = {
      inlineStatuses: getValue(transactionsStore.inlineStatuses),
      globalStatuses: getValue(transactionsStore.globalStatuses),
    };
    const args = '2';
    transactionsStore.switchStatusLocation(args);

    expect({
      prevValue,
      args,
      result: {
        inlineStatuses: getValue(transactionsStore.inlineStatuses),
        globalStatuses: getValue(transactionsStore.globalStatuses),
      },
    }).toMatchSnapshot();
  });

  it('should move status from global to inline', () => {
    const prevValue = {
      inlineStatuses: getValue(transactionsStore.inlineStatuses),
      globalStatuses: getValue(transactionsStore.globalStatuses),
    };
    const args = '3';
    transactionsStore.switchStatusLocation(args);

    expect({
      prevValue,
      args,
      result: {
        inlineStatuses: getValue(transactionsStore.inlineStatuses),
        globalStatuses: getValue(transactionsStore.globalStatuses),
      },
    }).toMatchSnapshot();
  });
});

describe('displayStatusInline', () => {
  let transactionsStore: TransactionsStore;

  beforeEach(() => {
    transactionsStore = new TransactionsStore();
    transactionsStore.displayStatusInline('1');
    transactionsStore.displayStatusInline('2');
  });

  it('should add new status id into inlineStatuses queue and keep old statuses', () => {
    const prevValue = getValue(transactionsStore.inlineStatuses);
    const args = '3';
    transactionsStore.displayStatusInline(args);

    expect({
      prevValue,
      args,
      result: getValue(transactionsStore.inlineStatuses),
    }).toMatchSnapshot();
  });

  it('should not duplicate items in inlineStatuses', () => {
    const prevValue = getValue(transactionsStore.inlineStatuses);
    const args = '2';
    transactionsStore.displayStatusInline(args);

    expect({
      prevValue,
      args,
      result: getValue(transactionsStore.inlineStatuses),
    }).toMatchSnapshot();
  });

  it('should remove status from globalStatuses if it exists there', () => {
    const args = '3';

    transactionsStore.displayStatusGlobal(args);
    const prevValue = getValue(transactionsStore.globalStatuses);
    transactionsStore.displayStatusInline(args);

    expect({
      prevValue,
      args,
      result: getValue(transactionsStore.globalStatuses),
    }).toMatchSnapshot();
  });
});

describe('displayStatusGlobal', () => {
  let transactionsStore: TransactionsStore;

  beforeEach(() => {
    transactionsStore = new TransactionsStore();
    transactionsStore.displayStatusGlobal('1');
    transactionsStore.displayStatusGlobal('2');
  });

  it('should add new status id into globalStatuses queue and keep old statuses', () => {
    const prevValue = getValue(transactionsStore.globalStatuses);
    const args = '3';

    transactionsStore.displayStatusGlobal(args);
    expect({
      prevValue,
      args,
      result: getValue(transactionsStore.globalStatuses),
    }).toMatchSnapshot();
  });

  it('should not duplicate items in globalStatuses', () => {
    const prevValue = getValue(transactionsStore.globalStatuses);
    const args = '2';

    transactionsStore.displayStatusGlobal(args);
    expect({
      prevValue,
      args,
      result: getValue(transactionsStore.globalStatuses),
    }).toMatchSnapshot();
  });

  it('should remove status from inlineStatuses if it exists there', () => {
    const args = '3';
    transactionsStore.displayStatusInline(args);
    const prevValue = getValue(transactionsStore.inlineStatuses);
    transactionsStore.displayStatusGlobal(args);

    expect({
      prevValue,
      args,
      result: getValue(transactionsStore.inlineStatuses),
    }).toMatchSnapshot();
  });
});

describe('hideStatus', () => {
  let transactionsStore: TransactionsStore;

  beforeEach(() => {
    transactionsStore = new TransactionsStore();
    transactionsStore.displayStatusGlobal('1');
    transactionsStore.displayStatusGlobal('2');
    transactionsStore.displayStatusInline('3');
    transactionsStore.addTransactionPayload('1', 'txPayload' as unknown as TransactionPayload);
    transactionsStore.addTransactionPayload('2', 'txPayload' as unknown as TransactionPayload);
    transactionsStore.addTransactionPayload('3', 'txPayload' as unknown as TransactionPayload);
    transactionsStore.addTxStatus('1', { txHash: '1', result: success(receiptMock) });
    transactionsStore.addTxStatus('2', { txHash: '2', result: success(receiptMock) });
    transactionsStore.addTxStatus('3', { txHash: '3', result: success(receiptMock) });
  });

  it('should remove tx ID from globalStatuses', () => {
    const prevValue = getValue(transactionsStore.globalStatuses);
    const args = '2';

    transactionsStore.hideStatus(args);
    expect({
      prevValue,
      args,
      result: getValue(transactionsStore.globalStatuses),
    }).toMatchSnapshot();
  });

  it('should remove tx ID from inlineStatuses', () => {
    const prevValue = getValue(transactionsStore.inlineStatuses);
    const args = '3';

    transactionsStore.hideStatus('3');
    expect({
      prevValue,
      args,
      result: getValue(transactionsStore.inlineStatuses),
    }).toMatchSnapshot();
  });

  it('should remove tx status', () => {
    const prevValue = getValue(transactionsStore.txStatuses);
    const args = '2';

    transactionsStore.hideStatus(args);
    expect({
      prevValue,
      args,
      result: getValue(transactionsStore.txStatuses),
    }).toMatchSnapshot();
  });

  it('should remove tx payload', () => {
    const prevValue = getValue(transactionsStore.transactionPayloads);
    const args = '2';

    transactionsStore.hideStatus(args);
    expect({
      prevValue,
      args,
      result: getValue(transactionsStore.transactionPayloads),
    }).toMatchSnapshot();
  });
});
