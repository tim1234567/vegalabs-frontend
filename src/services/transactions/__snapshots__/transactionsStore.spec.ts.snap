// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`addTransactionPayload should create new payload record and keep old payloads 1`] = `
Object {
  "args": Array [
    "3",
    "txPayload",
  ],
  "prevValue": Object {
    "1": "txPayload",
    "2": "txPayload",
  },
  "result": Object {
    "1": "txPayload",
    "2": "txPayload",
    "3": "txPayload",
  },
}
`;

exports[`addTransactionPayload should replace payload for existing transaction ID 1`] = `
Object {
  "args": Array [
    "1",
    Object {
      "txStatusContent": Object {
        "network": "polygon",
        "title": Object {
          "long": "111",
          "short": "1",
        },
      },
    },
  ],
  "prevValue": Object {
    "1": "txPayload",
    "2": "txPayload",
  },
  "result": Object {
    "1": Object {
      "txStatusContent": Object {
        "network": "polygon",
        "title": Object {
          "long": "111",
          "short": "1",
        },
      },
    },
    "2": "txPayload",
  },
}
`;

exports[`addTxStatus should create new status record and keep old statuses 1`] = `
Object {
  "args": Array [
    "3",
    Object {
      "result": Loading {
        "value": Object {
          "tag": "LOADING",
        },
      },
      "txHash": "3",
    },
  ],
  "prevValue": Object {
    "1": Object {
      "result": Loading {
        "value": Object {
          "tag": "LOADING",
        },
      },
      "txHash": "1",
    },
    "2": Object {
      "result": Loading {
        "value": Object {
          "tag": "LOADING",
        },
      },
      "txHash": "2",
    },
  },
  "result": Object {
    "1": Object {
      "result": Loading {
        "value": Object {
          "tag": "LOADING",
        },
      },
      "txHash": "1",
    },
    "2": Object {
      "result": Loading {
        "value": Object {
          "tag": "LOADING",
        },
      },
      "txHash": "2",
    },
    "3": Object {
      "result": Loading {
        "value": Object {
          "tag": "LOADING",
        },
      },
      "txHash": "3",
    },
  },
}
`;

exports[`addTxStatus should replace status for existing transaction ID 1`] = `
Object {
  "args": Array [
    "1",
    Object {
      "result": Success {
        "value": Object {
          "data": "ReceiptMock",
          "tag": "SUCCESS",
        },
      },
      "txHash": "111",
    },
  ],
  "prevValue": Object {
    "1": Object {
      "result": Loading {
        "value": Object {
          "tag": "LOADING",
        },
      },
      "txHash": "1",
    },
    "2": Object {
      "result": Loading {
        "value": Object {
          "tag": "LOADING",
        },
      },
      "txHash": "2",
    },
  },
  "result": Object {
    "1": Object {
      "result": Success {
        "value": Object {
          "data": "ReceiptMock",
          "tag": "SUCCESS",
        },
      },
      "txHash": "111",
    },
    "2": Object {
      "result": Loading {
        "value": Object {
          "tag": "LOADING",
        },
      },
      "txHash": "2",
    },
  },
}
`;

exports[`displayStatusGlobal should add new status id into globalStatuses queue and keep old statuses 1`] = `
Object {
  "args": "3",
  "prevValue": Array [
    "1",
    "2",
  ],
  "result": Array [
    "1",
    "2",
    "3",
  ],
}
`;

exports[`displayStatusGlobal should not duplicate items in globalStatuses 1`] = `
Object {
  "args": "2",
  "prevValue": Array [
    "1",
    "2",
  ],
  "result": Array [
    "1",
    "2",
  ],
}
`;

exports[`displayStatusGlobal should remove status from inlineStatuses if it exists there 1`] = `
Object {
  "args": "3",
  "prevValue": Array [
    "3",
  ],
  "result": Array [],
}
`;

exports[`displayStatusInline should add new status id into inlineStatuses queue and keep old statuses 1`] = `
Object {
  "args": "3",
  "prevValue": Array [
    "1",
    "2",
  ],
  "result": Array [
    "1",
    "2",
    "3",
  ],
}
`;

exports[`displayStatusInline should not duplicate items in inlineStatuses 1`] = `
Object {
  "args": "2",
  "prevValue": Array [
    "1",
    "2",
  ],
  "result": Array [
    "1",
    "2",
  ],
}
`;

exports[`displayStatusInline should remove status from globalStatuses if it exists there 1`] = `
Object {
  "args": "3",
  "prevValue": Array [
    "3",
  ],
  "result": Array [],
}
`;

exports[`hideStatus should remove tx ID from globalStatuses 1`] = `
Object {
  "args": "2",
  "prevValue": Array [
    "1",
    "2",
  ],
  "result": Array [
    "1",
  ],
}
`;

exports[`hideStatus should remove tx ID from inlineStatuses 1`] = `
Object {
  "args": "3",
  "prevValue": Array [
    "3",
  ],
  "result": Array [],
}
`;

exports[`hideStatus should remove tx payload 1`] = `
Object {
  "args": "2",
  "prevValue": Object {
    "1": "txPayload",
    "2": "txPayload",
    "3": "txPayload",
  },
  "result": Object {
    "1": "txPayload",
    "3": "txPayload",
  },
}
`;

exports[`hideStatus should remove tx status 1`] = `
Object {
  "args": "2",
  "prevValue": Object {
    "1": Object {
      "result": Success {
        "value": Object {
          "data": "ReceiptMock",
          "tag": "SUCCESS",
        },
      },
      "txHash": "1",
    },
    "2": Object {
      "result": Success {
        "value": Object {
          "data": "ReceiptMock",
          "tag": "SUCCESS",
        },
      },
      "txHash": "2",
    },
    "3": Object {
      "result": Success {
        "value": Object {
          "data": "ReceiptMock",
          "tag": "SUCCESS",
        },
      },
      "txHash": "3",
    },
  },
  "result": Object {
    "1": Object {
      "result": Success {
        "value": Object {
          "data": "ReceiptMock",
          "tag": "SUCCESS",
        },
      },
      "txHash": "1",
    },
    "3": Object {
      "result": Success {
        "value": Object {
          "data": "ReceiptMock",
          "tag": "SUCCESS",
        },
      },
      "txHash": "3",
    },
  },
}
`;

exports[`switchStatusLocation should keep old statuses if switched status is not found 1`] = `
Object {
  "args": "5",
  "prevValue": Object {
    "globalStatuses": Array [
      "3",
      "4",
    ],
    "inlineStatuses": Array [
      "1",
      "2",
    ],
  },
  "result": Object {
    "globalStatuses": Array [
      "3",
      "4",
    ],
    "inlineStatuses": Array [
      "1",
      "2",
    ],
  },
}
`;

exports[`switchStatusLocation should move status from global to inline 1`] = `
Object {
  "args": "3",
  "prevValue": Object {
    "globalStatuses": Array [
      "3",
      "4",
    ],
    "inlineStatuses": Array [
      "1",
      "2",
    ],
  },
  "result": Object {
    "globalStatuses": Array [
      "4",
    ],
    "inlineStatuses": Array [
      "1",
      "2",
      "3",
    ],
  },
}
`;

exports[`switchStatusLocation should move status from inline to global 1`] = `
Object {
  "args": "2",
  "prevValue": Object {
    "globalStatuses": Array [
      "3",
      "4",
    ],
    "inlineStatuses": Array [
      "1",
      "2",
    ],
  },
  "result": Object {
    "globalStatuses": Array [
      "3",
      "4",
      "2",
    ],
    "inlineStatuses": Array [
      "1",
    ],
  },
}
`;

exports[`updateTxStatus should not change statuses if they do not contain the transaction ID 1`] = `
Object {
  "args": Array [
    "3",
    Object {
      "txHash": "3",
    },
  ],
  "prevValue": Object {
    "1": Object {
      "result": Loading {
        "value": Object {
          "tag": "LOADING",
        },
      },
      "txHash": "1",
    },
    "2": Object {
      "result": Loading {
        "value": Object {
          "tag": "LOADING",
        },
      },
      "txHash": "2",
    },
  },
  "result": Object {
    "1": Object {
      "result": Loading {
        "value": Object {
          "tag": "LOADING",
        },
      },
      "txHash": "1",
    },
    "2": Object {
      "result": Loading {
        "value": Object {
          "tag": "LOADING",
        },
      },
      "txHash": "2",
    },
  },
}
`;

exports[`updateTxStatus should partially update status for existing transaction ID 1`] = `
Object {
  "args": Array [
    "1",
    Object {
      "txHash": "111",
    },
  ],
  "prevValue": Object {
    "1": Object {
      "result": Loading {
        "value": Object {
          "tag": "LOADING",
        },
      },
      "txHash": "1",
    },
    "2": Object {
      "result": Loading {
        "value": Object {
          "tag": "LOADING",
        },
      },
      "txHash": "2",
    },
  },
  "result": Object {
    "1": Object {
      "result": Loading {
        "value": Object {
          "tag": "LOADING",
        },
      },
      "txHash": "111",
    },
    "2": Object {
      "result": Loading {
        "value": Object {
          "tag": "LOADING",
        },
      },
      "txHash": "2",
    },
  },
}
`;
